oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,569 @@
+    pub fn len(&self) -> u64 {
+        self.marker_positions[4]
+    }
+
+    /// Determine if the number of samples in the population are empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Return the marker heights.
+    pub fn marker_heights(&self) -> [f64; FILLED_MARKER_LEN] {
+        self.marker_heights
+    }
+
+    /// Return the marker positions.
+    pub fn marker_positions(&self) -> [u64; FILLED_MARKER_LEN] {
+        self.marker_positions
+    }
+
+    /// Return the desired marker positions.
+    pub fn desired_marker_positions(&self) -> [f64; FILLED_MARKER_LEN] {
+        self.desired_marker_positions
+    }
+
+    /// Estimate the p-quantile of the population.
+    ///
+    /// This is step B.4 in the P² algorithm.
+    ///
+    /// Returns a result containing the estimated p-quantile or an error.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`QuantileError::InsufficientSampleSize`] if the sample size
+    /// is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oximeter::Quantile;
+    /// let mut q = Quantile::new(0.5).unwrap();
+    /// for o in 1..=100 {
+    ///    q.append(o).unwrap();
+    /// }
+    /// assert_eq!(q.estimate().unwrap(), 50.0);
+    /// ```
+    pub fn estimate(&self) -> Result<f64, QuantileError> {
+        if self.is_empty() {
+            return Err(QuantileError::InsufficientSampleSize);
+        }
+
+        if self.len() >= FILLED_MARKER_LEN as u64 {
+            return Ok(self.marker_heights[2]);
+        }
+
+        // Try to find an index in heights that is correlated with the p value
+        // when we have less than 5 samples, but more than 0.
+        let mut heights = self.marker_heights;
+        float_ord::sort(&mut heights);
+        let idx = (heights.len() as f64 - 1.) * self.p();
+        return Ok(heights[idx.round() as usize]);
+    }
+
+    /// Append a value/observation to the population and adjust the heights.
+    ///
+    /// This comprises steps B.1, B.2, B.3 (adjust heights) in the P² algorithm,
+    /// including finding the cell k containing the input value and updating the
+    /// current and desired marker positions.
+    ///
+    /// Returns an empty result or an error.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`QuantileError::NonFiniteValue`] if the value is not finite
+    /// when casting to a float.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oximeter::Quantile;
+    /// let mut q = Quantile::new(0.9).unwrap();
+    /// q.append(10).unwrap();
+    /// assert_eq!(q.len(), 1);
+    /// ```
+    pub fn append<T>(&mut self, value: T) -> Result<(), QuantileError>
+    where
+        T: HistogramSupport,
+    {
+        if !value.is_finite() {
+            return Err(QuantileError::NonFiniteValue);
+        }
+        // We've already checked that the value is finite.
+        let value_f = value.to_f64().unwrap();
+
+        // if !self.is_filled {
```

Comment:
Is this dead code, or a comment? I like the method name or field, if you want to keep it. It does make clear the intent much better than the check against the const.

Reply:
Ah, dead code. Will remove.